### PR TITLE
Baggage propagator first implementation

### LINKED ISSUE #1530: CorrelationsContext does not work from the client
**Is your feature request related to a problem? Please describe.**
I have come from OpenTracing and I would like to use OpenTelemetry in a complete replacement. I am advocating it for use in passing feature flag state from clients to servers so they propagate through the infrastructure (which we do for OpenTracing). However, with the Java implementation (instrumentation) this simply doesn't work. 

This for example with the Jaeger libraries and OpenTracing works fine, baggage appears:

curl -v -H "jaeger-debug-id: 11" -H "uberctx-fhub.FEATURE_TITLE_TO_UPPERCASE: true" -H 'content-type: application/json' http://localhost:8099/todo/list

With OpenTelemetry replacing it, this doesn't work:

curl -v -H 'traceparent: 00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01' -H 'otcorrelations: FEATURE_TITLE_TO_UPPERCASE=true'  -H 'content-type: application/json' http://localhost:8099/todo/list

I can see the span and parent and so forth all turning up, but the CorrelationContext is always empty. 

OpenTelemetry.getCorrelationContextManager().getCurrentContext().getEntries()

**Describe the solution you'd like**
Either this to work or some indication as to why it isn't working or I have it completely wrong and it will never work. 

**Describe alternatives you've considered**
I'm not aware apart from sticking with OpenTracing what other options there are.

**Additional context**

### PR DESCRIPTION
**Pull request by jkwatson**
This resolves #1530

### CODE REVIEW

File: api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  private final BaggageManager baggageManager;
+
+  // visible for testing
+  W3CBaggagePropagator(BaggageManager baggageManager) {
+    this.baggageManager = baggageManager;
+  }
+
+  /**
+   * Singleton instance of the W3C Baggage Propagator. Uses the {@link BaggageManager} from the
```
[the remark this answers is not in the document]
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java:39]: I'm really hoping we can completely get rid of the BaggageManager once https://github.com/open-telemetry/opentelemetry-specification/issues/1070 is resolved. If the Baggage in the API isn't a no-op, then we don't have to worry about this thing.

File: api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  @Override
+  public <C> Context extract(Context context, C carrier, Getter<C> getter) {
+    String baggageHeader = getter.get(carrier, FIELD);
+    if (baggageHeader == null) {
+      return context;
+    }
+    if (baggageHeader.isEmpty()) {
+      return BaggageUtils.withBaggage(EmptyBaggage.getInstance(), context);
```
[the remark this answers is not in the document]
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java:79]: It's not clear to me, either, if no baggage and empty baggage are semantically different.

File: api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+  @Override
+  public List<String> fields() {
+    return FIELDS;
+  }
+
+  @Override
+  public <C> void inject(Context context, C carrier, Setter<C> setter) {
+    Baggage baggage = BaggageUtils.getBaggage(context);
+    if (baggage == null) {
```
[the remark this answers is not in the document]
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java:54]: agreed. Assuming no baggage and empty baggage are semantic identical.

File: api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (anuraaga) [reply api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java:39]: Me too :)

File: api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    try {
+      extractEntries(baggageHeader, baggageBuilder);
+    } catch (Exception e) {
+      return BaggageUtils.withBaggage(EmptyBaggage.getInstance(), context);
+    }
+    return BaggageUtils.withBaggage(baggageBuilder.build(), context);
+  }
+
+  @SuppressWarnings("StringSplitter")
+  private static void extractEntries(String baggageHeader, Baggage.Builder baggageBuilder) {
```
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java:92]: Random noise from me - not sure what the difference between baggage and tracestate is, they're both propagated K-V pairs it seems. Anyways ;)

**Review by anuraaga — APPROVED** [review]

**Comment by anuraaga** [comment]
Let's get this in and fix forward after clarifying the spec more, thanks!

File: api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java (hunk 4 above)
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/baggage/propagation/W3CBaggagePropagator.java:92]: My feeling is: TraceState is a fixed format, semantically well-defined spec. Baggage is the wild west free-for-all for whatever users want to do.

**Pull request merged**